Bundle Order Import

braidn opened the pull request:
Order Import working on Bundled Products.

_Aside_:
This will shrink substantially when #24 is merged with master.

braidn commented:
The more that I play here @cabouffard @DanielWright and @bryanmtl it feels just not right to intermingle things from Product Assembly and this gem. That glue code seems like it should be in the host app right(Glossier)? It should be our job to establish some hooks to make this glue code easier to write? 

For now, should we ignore this feeling and build this feature in the gem(here) itself?

bryanmtl commented:
The product assembly gem is going to be a part of Glossier's stack for the foreseeable future. I would prefer to have the logic encapsulated in this Gem (and then removed when/if it's open-sourced). That said, I could also be persuaded to include it in the host app.

braidn commented:
Bundle Products take 1 ✅

braidn commented:
@cabouffard currently this uses the first part of the bundled sku as the lookup for the variant to add. Is there going to be a single product with many bundles or each bundle difference(colors/products/etc) will be 'under' a different product/master_variant?

thisiscab commented:
I am unsure I understand your question.

On shopify, there will be a product named `Stage 2`. That product is considered as a Bundle in the sense the product will contains all the possible permutation as the variants. 

So that `Stage 2` product may contains 60 variants.  Each of those variants will follow the convention we have set as the `sku` name. 

Does that answer your question?

braidn commented:
Kinda... this is what currently happens:
1. We read the first part of the SKU: in this case "PHASE2"
2. I look up a bundle on this.
3. I process the bundle as a line item, etc.

What I think I need to do is:
- [x] Read the first part of the SKU for "PHASE2"
- [x] Look up the bundle.
- [x] Use the remaining _Shopify Sku_ to figure out the specific variant with the right part combination.
- [x] Process this bundle/variant as the line_item variant.

thisiscab commented:
That looks like that could totally work :)

braidn committed:
Process Part Line Items From POS Orders

If a POS order has a bundle as a line item, the corresponding order will
be generated properly now in Solidus. This really means that the line
item that has the assembled product will have the correct corresponding
part_line_items.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
Line Item Bundled Products On Import

When a POS order includes bundled products, it should process the line
item as a bundled product, pulling the correct variant and setting the
price paid for the bundle.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn commented:
ok @cabouffard @bryanmtl and @DanielWright this branch/pr is now open for code review after pulling in the much larger order importer. Thank you

_Context_:
This PR allows the order import from #24 to consume and produce orders with line items that are bundled products. It associates the correct bundled product to the line item variant and pulls in all 'selected' variations of the variant into the part_line_items. This is a ridiculously named join table for parts and line items.

braidn commented:
Re-ping here for CR @cabouffard @bryanmtl and @DanielWright . Changes here are very minimal.

thisiscab commented on lib/solidus_retail/order/generate_pos_order.rb:
```diff
@@ -45,17 +45,37 @@ def create_order
 
```
I didn't realize in your last PR that everything was put inside the `lib` folder instead of the `app` folder. 

I would like if the code would follow the directory structure that is on master. This may be outside to scope of this PR but I wanted to flag it at least.

braidn replied:
Why would this need to be under the app directory? This doesn't really rely on Rails at all

thisiscab replied:
I'd like to follow some kind of convention instead of having the code at two different places. 

I'd like to hear your suggestions about how we can have a middle ground here. 
Here are mines:
- One suggestion would be that I could move non-rails related stuff to the lib folder
- Another would be that you move your stuff into the app folder

braidn replied:
I can totally move my code into App/ . It's generally a 'best' practice to avoid this because it's much quicker to test. However, it's only a lazy developer best practice because we are attempting to speed up the red/green/refactor loop that comes with TDD

thisiscab commented on lib/solidus_retail/order/generate_pos_order.rb:
```diff
@@ -45,17 +45,37 @@ def create_order
+
+      def add_line_item_parts(item, line_item)
+        Spree::Variant.where(sku: item.sku.split('/')[1..3]).each do |part_variant|
+          line_item.part_line_items.create( variant: part_variant, quantity: 1, line_item: line_item )
```
Extra space between `create(` and `variant`. Same for the end parenthesis.

braidn replied:
Will fix!

thisiscab commented on spec/solidus_retail/order/generate_pos_order_spec.rb:
```diff
@@ -33,6 +33,36 @@
       subject
       expect(last_order).to be_complete
     end
+
```
Are those considered as request specs?

braidn replied:
No, this is more of a straight unit test

thisiscab replied:
Then if it's unit test wouldn't be better off to mock everything and maybe use the helper method that I have provided in order to create double object instead of relying in creating Solidus objects?

braidn replied:
Everything is mocked besides the data right? We care about the end result being a complete and whole Solidus Order right? Not a bunch of hypothetically mocked items in the Order correct? That seems brittle to me

braidn committed:
Remove Extra Space In Parens

Weird extra space in parens removed. Likely created from changing the
entire contents of of the parens. Fulfills code review comment linked
below.

GitHub:
https://github.com/glossier/solidus_retail/pull/36#pullrequestreview-2019026

thisiscab commented:
By the way you got my pair of 👀 . This has been CRed :)

braidn commented:
@cabouffard I think the only thing that needs to be resolved is if you want this under the app directory?

thisiscab commented:
I see your point that the current code should have been maybe put into the `lib` folder but all the current code resides in the `app` folder.

I think it would make more sense for your `lib` code to be moved into the `app` folder. If you think it would be a better decision to more the `app` code into the `lib` code I wouldn't argue, I just want to make the code all in the same place if possible.

braidn commented:
Deal! Will move.

braidn committed:
Change The Directory For POS Order Class

Move everything under the App directory over the lib directory for the
pos order generator. This also changes the module structure which means
that the call has been changed to the rake task.

GitHub:
https://github.com/glossier/solidus_retail/pull/36#issuecomment-250498651

braidn commented:
💥 does is this the dir change you were envisioning @cabouffard ?

thisiscab commented:
That works
